### PR TITLE
108 check change units in p foresce lc per catchment reclass cat$total prms area

### PR DESCRIPTION
**Pull request by msleckman**
Added new functions in `2_process/src/area_diff_fix.R` to [1] check the area difference between gpkg catchments (`p1_PRMS_catchments_edits_sf`) and comid area aggregation in nlcd lc dataset using a specified km2 diff threshold; [2] Build a new shpfile with nhd-based catchments for the PRMS_segid that require new catchments polygons. 

Much of the updates are also created in `2_process.R`, where important new targets are added: 

`p2_PRMS_segid_special_handling_list` : list of PRMS_segid requiring new catchment polygon assignment w/ nhd.
`p2_nhd_catchments_dissolved_sf` : new sf object of nhd catchments dissolved to PRMS scale.
`p2_FORESCE_LC_per_catchment_fixed` : extraction of LC rasters from FORESCE lulc model using new nhd-based catchment shp and proportion calculation 
`p2_FORESCE_LC_per_catchment_fixed_reclass_cat` : reclassification of proportions for fixed PRMS_segid. (this is not done earlier because original reclassed LC proportion datasets are still scaled at hru level). 

Utimately the fixed and original lc proportions datasets for FORESCE are rebinded into target `p2_FORESCE_LC_per_catchment_reclass_cat`. 

Please note. These numerous targets were added to `2_process.R` for sake of time. If time allows, process this catchment fix in a separately sourced R script/function. 

New scatterplot after fixing (code below): 
image

@jds485 , I have not updated target `p2_FORESCE_LC_per_catchment_reclass_tot` - which uses the recursive fun - based on these changes. I'll let you take that over. 

The Rdsalt target also uses the raster extraction and proportion calc function using the gpkg catchment target (p1_catchments_edited_sf). This polygon fix was not made for the rdsalt target because this target does not get appended to another nhd-based dataset such as nlcd. We should decide whether we want to stick with the gpkg shp as the catchment boundary for rd salt, or if we want to dissolve all nhd catchments to PRMS scale for all PRMS_segid if we want more granular rdsalt measurements. 

___________
```
library(dplyr)
library(tidyverse)
library(nhdplusTools)
library(mapview)

# Comparing PRMS NLCD proportions dataset wih FORESCE
targets::tar_make(p2_PRMS_NLCD_lc_proportions_reclass_cat)
targets::tar_make(p2_FORESCE_LC_per_catchment_reclass_cat)
targets::tar_load(p2_PRMS_NLCD_lc_proportions_reclass_cat)
targets::tar_load(p2_FORESCE_LC_per_catchment_reclass_cat)

# subset/filter and mutate. We choose year 2000 for FORESCE and 2001 for NLCD
##2000
l1 <- p2_FORESCE_LC_per_catchment_reclass_cat[[7]] %>%
  #  select(PRMS_segid, Year, everything(), -hru_segment) %>% 
 # mutate(total_PRMS_area_km2 = total_PRMS_area/10^6) %>%
  select(PRMS_segid, total_PRMS_area_km2)
## 2001
l2 <- p2_PRMS_NLCD_lc_proportions_reclass_cat[[1]] %>%
#  rename(total_PRMS_area_km2 = AREASQKM_PRMS) %>% 
  # select(PRMS_segid, Year, everything(), -AREASQKM_PRMS) %>% 
  select(PRMS_segid, AREASQKM_PRMS)

# Merge l1 and l2 and calculate difference for each PRMS SEGMENT
df <- l1 %>% left_join(l2, by = 'PRMS_segid') %>% 
  rename(PRMS_area_km2_FOR = total_PRMS_area_km2, PRMS_area_km2_NLCD = AREASQKM_PRMS) %>% 
  mutate(ab_diff = abs(PRMS_area_km2_FOR - PRMS_area_km2_NLCD)) %>%
  mutate(ab_diff_categorized = factor(case_when(ab_diff > 50 ~ '>50 km2',
                              ab_diff > 10 ~ '>10 km2',
                              ab_diff > 5 ~ '>5 km2',
                              ab_diff >1 ~ '>1 km2',
                              TRUE ~ '< 1 km2'),
                              levels = c('< 1 km2', '>1 km2', '>5 km2', '>10 km2', '>50 km2'))) 

# Plotting
scatterplot1 <- ggplot(df, aes(x = PRMS_area_km2_FOR, y = PRMS_area_km2_NLCD))+
  geom_point(aes(color = ab_diff_categorized))+theme_bw()+
  labs(caption = 'Scatter plot of PRMS area in the NLCD dataset versus FORESCE.\n Point colored by their difference categories')+
  theme(plot.caption = element_text(hjust = 0.5))
scatterplot1  
 ```

### CODE REVIEW

File: 2_process/src/area_diff_fix.R [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    #assign area to reach 287_1 using reach length
+    mutate(areasqkm = case_when(PRMS_segid == '287_1' ~ (filter(prms_reaches_sf, 
+                                                                subsegid == '287_1') %>% 
+                                                           pull(subseglen))^2/10^6,
+                                TRUE ~ areasqkm))
+  
+  #assign geometry to reach 287_1 using a small buffer of 15 m to closely match approx. area
+  nhd_catchments$geometry[nhd_catchments$PRMS_segid == '287_1'] <- 
+    st_geometry(st_buffer(filter(prms_reaches_sf, subsegid == '287_1'), dist = 15))
+
```
[the remark this answers is not in the document]
Comment (jds485) [reply 2_process/src/area_diff_fix.R:67]: Noting that the FORE-SCE and NLCD proportions for 2000 and 2001, respectively, are very different for this reach. Added a task to issue #102 to revisit how the 2000 and 2001 data compare across all reaches.

NLCD:
`PRMS_segid AREASQKM_PRMS prop_lcClass_1 prop_lcClass_4 prop_lcClass_5 prop_lcClass_6 prop_lcClass_7
`
`
1      287_1      0.002304         0.0329         0.3623         0.4825         0.0059         0.0925
`
`
prop_lcClass_8 prop_lcClass_9 prop_lcClass_2 prop_lcClass_3 prop_lcClass_10 Year
`
`
1         0.0028         0.0011          0.009              0          0.0109 2001
`

FORE-SCE
`PRMS_segid hru_segment total_PRMS_area_km2 prop_lcClass_3 prop_lcClass_1 prop_lcClass_4 prop_lcClass_5
`
`
1      287_1         287         0.002286752              0            0.5              0              0
`
`
  prop_lcClass_7 prop_lcClass_8 prop_lcClass_9 prop_lcClass_10 prop_lcClass_6 prop_lcClass_2 Year
`
`
1            0.5              0              0               0              0              0 2000
`

**Review by jds485 — APPROVED** [review]
Looks good! You should check my edits to make a polygon for the reach without an NHD catchment polygon, but I think this can be merged!

File: 2_process/src/area_diff_fix.R [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  # join/filter comids polygons to PRMS_segid from initial list. Make into sf_object
+  nhd_catchments <- comids_PRMS_catchments_to_fix %>% 
+    left_join(shp_comids_PRMS_catchments_to_fix, by = c('comid' = 'featureid')) %>% 
+    ## turn to sf obj to then be able to dissolve 
+    sf::st_as_sf()
+  
+  ## Creating dissolved shp for PRMS_segid for different area
+  nhd_catchments_dissolved <- nhd_catchments %>% 
+    # ## 287_1 has no comid. removing this otherwise cannot transform to appropriate sf object
+    filter(PRMS_segid != '287_1') %>%
```
[the remark this answers is not in the document]
Comment (msleckman) [reply 2_process/src/area_diff_fix.R]: Is there another way to get the catchment of a comid? Or are we okay omitting 287_1

File: 2_process/src/area_diff_fix.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (msleckman) [reply 2_process/src/area_diff_fix.R:67]: I think this evaluation should occur in the combination step of NLCD and FORESCE.

File: 2_process/src/area_diff_fix.R [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  gpkg_df <- PRMS_shapefile  %>% st_drop_geometry() %>% group_by(PRMS_segid) %>% 
+    summarise(prms_gpkg_area_km2 = sum(hru_area_m2)/10^6)
+  
+  ##join gpkg catchments and nhd catchments, and calculate difference
+  nlcd_area <- nhd_catchment_areas %>%
+    select(PRMS_segid, comid, AREASQKM, TOTDASQKM, LENGTHKM) %>% 
+    rename(., nhd_cat_area_sqkm = AREASQKM, nhd_tot_area_sqkm = TOTDASQKM, nhd_len_km = LENGTHKM) %>% 
+    group_by(PRMS_segid) %>% 
+    summarize(nhd_cat_area_sqkm = sum(nhd_cat_area_sqkm),
+              nhd_tot_area_sqkm = sum(nhd_tot_area_sqkm),
```
[the remark this answers is not in the document]
Comment (msleckman) [reply 2_process/src/area_diff_fix.R:20]: yes made this update by removing from anything that relates to TOT from the code chunk.

**Comment by msleckman** [comment]
Read to be merged. Awaiting preceeding PRs before pushing.

**Comment by msleckman** [comment]
This is ready to merge following latest updates to main. 

Area check still the same graphic as above: 
image

**Pull request merged**